tiled/structures/awkward.py

```diff
@@ -0,0 +1,80 @@
+from dataclasses import dataclass
+
+import awkward
+
+
+@dataclass
+class AwkwardStructure:
+    length: int
+    form: dict
+
+    @classmethod
+    def from_json(cls, structure):
+        return cls(**structure)
+
+
+def project_form(form, form_keys_touched):
+    # See https://github.com/bluesky/tiled/issues/450
+    if isinstance(form, awkward.forms.RecordForm):
+        if form.fields is None:
+            original_fields = [None] * len(form.contents)
+        else:
+            original_fields = form.fields
+
+        fields = []
+        contents = []
+        for field, content in zip(original_fields, form.contents):
+            projected = project_form(content, form_keys_touched)
+            if projected is not None:
+                fields.append(field)
+                contents.append(content)
+
+        if form.fields is None:
+            fields = None
+
+        return form.copy(fields=fields, contents=contents)
+
+    elif isinstance(form, awkward.forms.UnionForm):
+        raise NotImplementedError
```

Comment:
Eventually, you'll be able to

```suggestion
    elif isinstance(form, awkward.forms.UnionForm):
        step1 = [project_form(x, form_keys_touched) for x in form.contents]
        step2 = [x for x in step1 if x is not None]
        if len(step2) == 0:
            return None
        elif len(step2) == 1:
            return step2[0]
        else:
            return awkward.forms.UnionForm.simplified(
                form.tags,
                form.index,
                step2,
                parameters=form.parameters,
                form_key=form.form_key,
            )
```

but the `awkward.forms.UnionForm.simplified` will drop the `form_keys` of all the `Content` instances in `step2`. We've been living with that problem for too long: I'm going to open a PR on Awkward to fix that, and when that's merged and in a new release, the above implementation will be 100% correct. For now, it will only work if `len(step2) < 2`.

Reply:
When/if scikit-hep/awkward#2661 goes into a release of Awkward Array, the above code will be correct.